API: sparse.linalg.svds: transition to rng (SPEC 7)

dschult opened the pull request:
Reference issue
Toward https://github.com/scipy/scipy/issues/21833

What does this implement/fix?
This allows for use of new keyword rng alongside random_state in sparse.linalg.svds to prepare for the deprecation and removal of random_state and its legacy behavior.

In short:

if you pass the RNG argument by position or keyword random_state, it should work however it did before
if you pass a value val by keyword rng, it should behave as if you had passed np.random.default_rng(val) into the function.

dschult committed:
use rng in sparse.linag. Specifically svds/svdp

j-bowhay commented on scipy/sparse/linalg/_eigen/_svds.py:
```diff
@@ -90,15 +90,16 @@ def _iv(A, k, ncv, tol, which, v0, maxiter,
+            return_singular, solver, rng)
 
 
+@_transition_to_rng("random_state")
```
Hopefully my counting is correct 
```suggestion
@_transition_to_rng("random_state", position_num=9)
```

dschult committed:
add position_num, lint, smoke-doc

dschult commented:
There's a doc_string test (line ~292 in `scipy/sparse/linalg/_eigen/_svds.py` ) that I've kept using kwarg `random_state=rng` because `rng` is set to be `RandomState(0)` and if I change that, the resulting values must be changed a few lines below.
```python
    >>> rng = np.random.RandomState(0)
    >>> X_dense = rng.random(size=(100, 100))
    >>> X_dense[:, 2 * np.arange(50)] = 0
    >>> X = sparse.csr_array(X_dense)
    >>> _, singular_values, _ = svds(X, k=5, random_state=rng)
    >>> print(singular_values)
    [ 4.3293...  4.4491...  4.5420...  4.5987... 35.2410...]

```
I could change the first line I've shown here to use `default_rng`, change the kwarg and change the values printed. But IIUC its better to leave `random_state` in order to minimize these changes. We'll deal with them upon removal. 

Do I understand correctly? Let me know if I should change the test from using RandomState to using default_rng.

mdhaber commented:
> Do I understand correctly? Let me know if I should change the test from using `RandomState` to using `default_rng`.

Please do change it, actually. It is good to do it now because we want new users who refer to the examples to use `rng`. And if it were not done now, it would need to be done in two versions when the decorator begins to emit warnings about use of `random_state`.

mdhaber commented on scipy/sparse/linalg/_eigen/tests/test_svds.py:
```diff
@@ -389,62 +385,62 @@ def test_svd_random_state(self):
         n = 100
         k = 5
 
         rng = np.random.default_rng(0)
         A = rng.random((n, n))
```
```suggestion
        rng_temp = np.random.default_rng(0)
        A = rng_temp.random((n, n))
```

dschult replied:
This involves collapsing two names onto one "rng".  Good catch! Sorry I didn't look for it.
And, this fix is also needed in the previous test.

But -- fixing this results in a TypeError when the parametrized `rng` is `RandomState(0)`.
So, I removed that parameter-value-mark just like was done with `test_randn` in the random_array PR.

mdhaber commented on scipy/sparse/linalg/_svdp.py:
```diff
@@ -79,10 +79,11 @@ def dtype(self):
             return self.A.matvec(np.zeros(self.A.shape[1])).dtype
 
 
+@_transition_to_rng("random_state", position_num=18)
```
Does this need the decorator to be applied? It's private, so it's OK to just change `random_state` to `rng` within.

dschult replied:
I wasn't sure how "private" it is because it is listed in `__all__`.  But it looks like no `__init__.py` files import it like that. So, in SciPy, it is only used by `svds` and in the associated tests for `propack`. I removed it here. A search for `"_svdp" site::https://github.com` shows only 5 results non of which are relevant. It's safe to say it isn't use elsewhere. I've updated accordingly here and in `_svds.py`.

mdhaber reviewed (commented):
Looks close. Besides the comments above, I'd just like to preserve at least one use of `random_state` as a sanity check that the decorator is doing its job.

dschult committed:
review comments and more stemming from those

dschult commented:
I think this addresses all points brought up -- and lead to a couple others which were easy to fix.

One last point is that **if** random_array PR gh-21888 gets merged first, this one could change quite a few `random_array` calls to use `rng=` instead of `random_state=`.  Both should work during the transition, so I think that's OK... but if we switch them now it makes the next transition simpler.

dschult committed:
make propack tests deterministic. remove random_state from _svdp

dschult committed:
make svds tests deterministic. allow RandomState in svdp

dschult commented:
I made all all svds tests deterministric and all propack tests that directly call `_svpd` deterministic.
The number of changes to make the tests deterministic is fairly large. 

Every function call to `_svdp` now specifies `rng`. It is still an "optional" argument though, so I added code to normalize the input `rng` inside `_svpd` unless it is a RandomState. I could put back the `check_random_state` code if that is preferred.

mdhaber commented on scipy/sparse/linalg/_eigen/tests/test_svds.py:
```diff
@@ -144,18 +144,16 @@ class SVDSCommonTests:
         else:
-            res = svds(A, k=1, which=which, solver=self.solver,
-                       random_state=0)
+            res = svds(A, k=1, which=which, solver=self.solver, random_state=0)
```
```suggestion
            # preserve a use of legacy keyword `random_state` during SPEC 7 transition
            res = svds(A, k=1, which=which, solver=self.solver, random_state=0)
```

mdhaber replied:
Wherever it is, it has been suggested that we leave a comment so it's clear that it's intentional. It appears to be here, do I suggested adding the comment.

dschult replied:
I put the same comment that you suggested in the other location (line 178).

mdhaber commented on scipy/sparse/linalg/_svdp.py:
```diff
@@ -174,7 +167,8 @@ def _svdp(A, k, which='LM', irl_mode=True, kmax=None,
 
     """
-    random_state = check_random_state(random_state)
+    if not isinstance(rng, np.random.RandomState):
+        rng = np.random.default_rng(rng)
```
This looks like the code corresponding with:

> Every function call to `_svdp` now specifies rng. It is still an "optional" argument though, so I added code to normalize the input `rng` inside `_svpd` unless it is a `RandomState`. I could put back the `check_random_state` code if that is preferred.

I agree that all calls to `_svdp` specify `rng`, and for the most part they have already normalized the argument (except for a call with `rng=0` in `check_svdp` and `test_examples`). For that reason, what do you think about removing this normalization altogether, since we won't want this special case for `RandomState` to remain after the SPEC 7 transition is complete?

mdhaber commented on scipy/sparse/linalg/_eigen/_svds.py:
```diff
@@ -290,21 +285,21 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
     ever explicitly constructed.
 
     >>> rng = np.random.default_rng(102524723947864966825913730119128190974)
     >>> G = sparse.random_array((8, 9), density=0.5, random_state=rng)
```
Can we change this one, too?
```suggestion
    >>> G = sparse.random_array((8, 9), density=0.5, rng=rng)
```

mdhaber replied:
Oh, right. Doing too many of these at once!

mdhaber reviewed (approved):
This looks great! I've suggested marking the one remaining use of `random_state` in a test so it's obvious that it's not a mistake, and there are two other little things, but I think we can merge this shortly. I do think this will merge before gh-21888, if that's OK.

dschult committed:
adjust rng treatment for _svdp

dschult reviewed (commented):
I removed the rng normalization from `_svdp` and replaced it with a check that `rng` input is a `Generator` instance.   I updated the tests and I think we're all ready to go here.

I had to move the sanity check of `random_state=` kwarg to a different test that doesn't use `_svdp`. :}

dschult commented:
Oops.. somehow I didn't commit my last set of changes.
pushed before committing i guess.

`_svdp` should work without any normalization because everything that calls it has already normalized it.  That said, if someone has been using it on their own despite it being private, if/when they don't provide an `rng` arg is defaults to None, which creates a strange error message. So, I put code to check if `rng is None` and raise a ValueError if that happens. But it would only happen if someone if directly using the private function.

dschult committed:
remaining changes

mdhaber reviewed (approved):
> So, I put code to check if rng is None and raise a ValueError if that happens. But it would only happen if someone if directly using the private function.

OK. Note that it is not required to provide helpful error messages to users in private functions. That is why I have been advocating for removing it entirely. But if you think it might be helpful to developers in the future, this isn't causing problems, so I wouldn't ask for it to be changed now.

Let's see if tests come back green, and if so, I'll go ahead and merge.

Thanks @dschult!

ilayn commented:
I am bit confused with this one here, I am receiving some errors while trying to enable the rng in #22748 but the `check_random_state` is generating instances of `np.random.mtrand._rand`. Shouldn't it create `default_rng()` when `rng` argument is `None`?

ilayn commented:
True but if `rng=None` then it does not matter the type of the PRNG object we create since there is no argument given. So instead of `RandomState`, `check_random_state` can produce a `Generator` already. Or am I misunderstanding?

mdhaber commented:
It depends on if the global seed has been set. If so, it needs to use the global `RandomState` produced by `check_random_state`. If not, it is _ok_ to use a unseeded `RandomState` produced by `check_random_state` because that is interchangeable, as far as the function should be concerned, to an unseeded `Generator`.

ilayn commented:
Ah OK so this is from the `np.random.seed(int)` cases. Thanks.
